src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs

```diff
@@ -657,6 +657,102 @@ public async Task WebHostStopAsyncUsesDefaultTimeoutIfNoTokenProvided()
             }
         }
 
+        [Fact]
+        public async Task HostPropagatesExceptionsThrownWithBackgroundServiceExceptionBehaviorOfStopHost()
+        {
+            using IHost host = CreateBuilder()
+                .ConfigureServices(
+                    services =>
+                    {
+                        services.AddHostedService(_ => new AsyncThrowingService(Task.CompletedTask));
+                        services.Configure<HostOptions>(
+                            options =>
+                            options.BackgroundServiceExceptionBehavior =
+                                BackgroundServiceExceptionBehavior.StopHost);
+                    })
+                .Build();
+
+            var lifetime = host.Services.GetRequiredService<IHostApplicationLifetime>();
+            var wasStartedCalled = false;
+            lifetime.ApplicationStarted.Register(() => wasStartedCalled = true);
+            var wasStoppingCalled = false;
+            lifetime.ApplicationStopping.Register(() => wasStoppingCalled = true);
+
+            await host.StartAsync();
+
+            Assert.True(
+                wasStartedCalled,
+                $"The {nameof(AsyncThrowingService)} should have thrown, calling Host.StopAsync, started should not have been called.");
+            Assert.True(
+                wasStoppingCalled,
+                $"The {nameof(AsyncThrowingService)} should have thrown, calling Host.StopAsync, stopping should have been called.");
+        }
+
+        [Fact]
+        public async Task HostStopsApplicationWithOneBackgroundServiceErrorAndOthersWithoutError()
+        {
+            var executionCount = 0;
+            using IHost host = CreateBuilder()
+                .ConfigureServices(
+                    services =>
+                    {
+                        services.AddHostedService(_ => new TestBackgroundService(Task.Delay(200), i => executionCount = i));
```

Comment:
Tests that depend on timing (like `Task.Delay(200)`) fail quite often in CI since we run the test so many times on so many different platforms. Even a `0.01%` chance of failing still becomes noisy in our CI.

It is better to explicitly control the timing using Tasks, TaskCompletionSource, etc to ensure the ordering happens as you want.

Reply:
See the conversation at [https://github.com/dotnet/runtime/pull/44953#discussion_r527144606](https://github.com/dotnet/runtime/pull/44953#discussion_r527144606), for more infromation.

Reply:
Thank you for the callout on that, I'll update this to use explicitly controlled timing instead of the unreliable timing.